### ISSUE TITLE
Update window is not responsive after application installed until user restart his mac

### ISSUE THREAD

**Issue by aleksiy805** [sparkle-project/Sparkle#1757]
Hello to all! 
First of all, I want to thank the authors for their work!

I have a Cocoa app. I distribute it as .dmg or .app files using Apple Developer Id (I sign all framework, app, dmg).

The problem: in some cases, I didn't understand when and why. It was reproduced only twice on Big Sur. App becomes not responding on the usually Sparkle Update Screen with HTML release notes when how I understand WKWebView is used. It also reproduces on WKWebViews in other parts of the app. And in the end user just receives the message from OS 'Application is not responding' and needs only to Force Quit. User starts app again and receive the same result. And only after the Mac Restart everything becomes fine. However URLSessions work without problems. Problems are only with WKWebView. The app uses sandbox:

	<key>com.apple.security.app-sandbox</key>
	<true/>
	<key>com.apple.security.files.user-selected.read-write</key>
	<true/>
	<key>com.apple.security.network.client</key>
	<true/>
	<key>com.apple.security.network.server</key>
	<true/>

Sparkle framework 2.x version '0b9bacfd'

I couldn't describe steps to reproduce. People who met this problem didn't meet it again after the Mac Restart. I can only say that It's the first app installation and launching any number of times before the Mac restart and in quite rare cases. But it's very problematic.

I clearly understand that it's very small probability your framework and XPC Services do this. But I believe somebody met the same problem and know how to solve if.

Thank you for any help.

I attaches System Report after Mac OS closed the app.

[Report.zip](https://github.com/sparkle-project/Sparkle/files/5969150/Report.zip)

**Comment by zorgiepoo** [comment]
Ouch. WKWebView was implemented recently in 1.x and 2.x - we still have a fallback to legacy WebView in some cases but now I wonder if it should be configurable.. This issue seems odd and worth investigating if I'm reading it right:

```
appMac blocked by akd [26407] thread 0xf2a73:
10  -[SUWKWebView webView:decidePolicyForNavigationAction:decisionHandler:] + 302 (Sparkle + 30449) [0x102f0f6f1] 1-10
    10  WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction(WebKit::WebPageProxy&, WTF::Ref<API::NavigationAction, WTF::DumbPtrTraits<API::NavigationAction> >&&, WTF::Ref<WebKit::WebFramePolicyListenerProxy, WTF::DumbPtrTraits<WebKit::WebFramePolicyListenerProxy> >&&, API::Object*)::$_5::operator()(WKNavigationActionPolicy, objc_object*) + 469 (WebKit + 3266213) [0x7fff3c8f16a5] 1-10
    10  WebKit::SOAuthorizationCoordinator::tryAuthorize(WTF::Ref<API::NavigationAction, WTF::DumbPtrTraits<API::NavigationAction> >&&, WebKit::WebPageProxy&, WTF::Function<void (bool)>&&) + 52 (WebKit + 3785294) [0x7fff3c97024e] 1-10
        10  WebKit::SOAuthorizationCoordinator::canAuthorize(WTF::URL const&) const + 56 (WebKit + 3785228) [0x7fff3c97020c] 1-10
        10  +[SOAuthorization canPerformAuthorizationWithURL:responseCode:] + 59 (AppSSO + 6571) [0x7fff330f69ab] 1-10
            10  +[SOAuthorizationCore canPerformAuthorizationWithURL:responseCode:useInternalExtensions:] + 68 (AppSSOCore + 19414) [0x7fff2af3bbd6] 1-10
            10  +[SOAuthorizationCore _canPerformAuthorizationWithURL:responseCode:callerBundleIdentifier:useInternalExtensions:] + 100 (AppSSOCore + 9401) [0x7fff2af394b9] 1-10
                10  +[SOAuthorizationCore _doAKshouldProcessURL:] + 54 (AppSSOCore + 19568) [0x7fff2af3bc70] 1-10
                10  +[AKAuthorizationController shouldProcessURL:] + 93 (AuthKit + 29816) [0x7fff2a333478] 1-10
                    10  -[AKAuthorizationController _appleIDAuthorizationURLs] + 168 (AuthKit + 31851) [0x7fff2a333c6b] 1-10
                    10  _NSXPCDistantObjectSimpleMessageSend1 + 30 (Foundation + 151708) [0x7fff2121009c] 1-10
                        10  -[NSXPCConnection _sendSelector:withProxy:arg1:] + 122 (Foundation + 151835) [0x7fff2121011b] 1-10
                        10  -[NSXPCConnection _sendInvocation:orArguments:count:methodSignature:selector:withProxy:] + 3090 (Foundation + 118889) [0x7fff21208069] 1-10
                            10  __NSXPCCONNECTION_IS_WAITING_FOR_A_SYNCHRONOUS_REPLY__ + 9 (Foundation + 126953) [0x7fff21209fe9] 1-10
                            10  xpc_connection_send_message_with_reply_sync + 238 (libxpc.dylib + 45484) [0x7fff201091ac] 1-10
                                10  dispatch_mach_send_with_result_and_wait_for_reply + 50 (libdispatch.dylib + 105326) [0x7fff20232b6e] 1-10
                                10  _dispatch_mach_send_and_wait_for_reply + 518 (libdispatch.dylib + 104310) [0x7fff20232776] 1-10
                                    10  mach_msg_trap + 10 (libsystem_kernel.dylib + 3710) [0x7fff20392e7e] 1-10
                                    *10  ipc_mqueue_receive_continue + 0 (kernel + 573344) [0xffffff800029bfa0] (blocked by turnstile waiting for akd [26407] thread 0xf2a73 after 2 hops) 1-10

akd [26407] thread 0xf2a73 is blocked by akd [26407] thread 0xe9706:
Thread 0xf2a73    DispatchQueue 65    10 samples (1-10)    priority 47 (base 47)
<thread QoS user initiated (requested user initiated, requested override user initiated, promote user interactive, work queue override user initiated, workloop servicer override user interactive), process sudden termination dirty, process unclamped, process received importance donation from appMac [27215], IO tier 0>
10  start_wqthread + 15 (libsystem_pthread.dylib + 9319) [0x7fff203c3467] 1-10
10  _pthread_wqthread + 314 (libsystem_pthread.dylib + 13465) [0x7fff203c4499] 1-10
    10  _dispatch_workloop_worker_thread + 819 (libdispatch.dylib + 80989) [0x7fff2022cc5d] 1-10
    10  _dispatch_lane_invoke + 426 (libdispatch.dylib + 41214) [0x7fff202230fe] 1-10
        10  _dispatch_lane_serial_drain + 263 (libdispatch.dylib + 38055) [0x7fff202224a7] 1-10
        10  _dispatch_mach_invoke + 498 (libdispatch.dylib + 112056) [0x7fff202345b8] 1-10
            10  _dispatch_lane_serial_drain + 263 (libdispatch.dylib + 38055) [0x7fff202224a7] 1-10
            10  _dispatch_mach_msg_invoke + 441 (libdispatch.dylib + 109127) [0x7fff20233a47] 1-10
                10  _dispatch_client_callout4 + 9 (libdispatch.dylib + 14439) [0x7fff2021c867] 1-10
                10  _xpc_connection_mach_event + 935 (libxpc.dylib + 47772) [0x7fff20109a9c] 1-10
                    10  _xpc_connection_call_event_handler + 56 (libxpc.dylib + 52264) [0x7fff2010ac28] 1-10
                    10  message_handler + 206 (Foundation + 138569) [0x7fff2120cd49] 1-10
                        10  -[NSXPCConnection _decodeAndInvokeMessageWithEvent:flags:] + 2271 (Foundation + 435576) [0x7fff21255578] 1-10
                        10  __NSXPCCONNECTION_IS_CALLING_OUT_TO_EXPORTED_OBJECT_S1__ + 10 (Foundation + 783266) [0x7fff212aa3a2] 1-10
                            10  ??? (akd + 160081) [0x10b5f1151] 1-10
                            10  -[AKURLBag appleIDAuthorizationURLs] + 46 (AuthKit + 193298) [0x7fff2a35b312] 1-10
                                10  -[AKURLBag configurationAtKey:fromCache:] + 60 (AuthKit + 194256) [0x7fff2a35b6d0] 1-10
                                10  -[AKURLBag _configurationsFromCache:withError:] + 24 (AuthKit + 195019) [0x7fff2a35b9cb] 1-10
                                    10  -[AKURLBag _urlBagFromCache:withError:] + 189 (AuthKit + 195271) [0x7fff2a35bac7] 1-10
                                    10  ??? (akd + 399965) [0x10b62ba5d] 1-10
                                        10  ??? (akd + 400813) [0x10b62bdad] 1-10
                                        10  ??? (akd + 401627) [0x10b62c0db] 1-10
                                            10  ??? (akd + 405048) [0x10b62ce38] 1-10
                                            10  ??? (CFNetwork + 79248) [0x7fff24718590] 1-10
                                                10  objc_alloc_init + 33 (libobjc.A.dylib + 151688) [0x7fff20283088] 1-10
                                                10  object_cxxConstructFromClass + 92 (libobjc.A.dylib + 48858) [0x7fff20269eda] 1-10
                                                    10  ??? (CFNetwork + 2034325) [0x7fff248f5a95] 1-10
                                                    10  _dispatch_once_wait + 76 (libdispatch.dylib + 16170) [0x7fff2021cf2a] 1-10
                                                        10  <patched truncated backtrace> 1-10
                                                        10  __ulock_wait + 10 (libsystem_kernel.dylib + 9678) [0x7fff203945ce] 1-10
                                                            *10  ??? (kernel + 6747600) [0xffffff800087f5d0] (blocked by turnstile waiting for akd [26407] thread 0xe9706) 1-10

akd [26407] Thread 0xe9706 blocked by akd [26407] thread 0xe96f5:
Thread 0xe9706    10 samples (1-10)    priority 47 (base 47)
  <thread QoS user initiated (requested default, promote user interactive, workloop servicer override default), process sudden termination dirty, process unclamped, process received importance donation from appMac [27215], IO tier 0>
  10  start_wqthread + 15 (libsystem_pthread.dylib + 9319) [0x7fff203c3467] 1-10
    10  _pthread_wqthread + 314 (libsystem_pthread.dylib + 13465) [0x7fff203c4499] 1-10
      10  _dispatch_workloop_worker_thread + 819 (libdispatch.dylib + 80989) [0x7fff2022cc5d] 1-10
        10  _dispatch_lane_invoke + 426 (libdispatch.dylib + 41214) [0x7fff202230fe] 1-10
          10  _dispatch_lane_serial_drain + 606 (libdispatch.dylib + 38398) [0x7fff202225fe] 1-10
            10  _dispatch_client_callout + 8 (libdispatch.dylib + 14279) [0x7fff2021c7c7] 1-10
              10  _dispatch_call_block_and_release + 12 (libdispatch.dylib + 9693) [0x7fff2021b5dd] 1-10
                10  ??? (akd + 439096) [0x10b635338] 1-10
                  10  -[AKURLBag APSEnvironment] + 26 (AuthKit + 193439) [0x7fff2a35b39f] 1-10
                    10  -[AKURLBag _requestEnvironmentsWithError:] + 50 (AuthKit + 194904) [0x7fff2a35b958] 1-10
                      10  -[AKURLBag _urlBagFromCache:withError:] + 189 (AuthKit + 195271) [0x7fff2a35bac7] 1-10
                        10  ??? (akd + 399965) [0x10b62ba5d] 1-10
                          10  ??? (akd + 400813) [0x10b62bdad] 1-10
                            10  ??? (akd + 401627) [0x10b62c0db] 1-10
                              10  ??? (akd + 405048) [0x10b62ce38] 1-10
                                10  ??? (CFNetwork + 79248) [0x7fff24718590] 1-10
                                  10  objc_alloc_init + 67 (libobjc.A.dylib + 151722) [0x7fff202830aa] 1-10
                                    10  object_cxxConstructFromClass + 92 (libobjc.A.dylib + 48858) [0x7fff20269eda] 1-10
                                      10  ??? (CFNetwork + 2034325) [0x7fff248f5a95] 1-10
                                        10  _dispatch_once_callout + 20 (libdispatch.dylib + 18795) [0x7fff2021d96b] 1-10
                                          10  _dispatch_client_callout + 8 (libdispatch.dylib + 14279) [0x7fff2021c7c7] 1-10
                                            10  ??? (CFNetwork + 1406024) [0x7fff2485c448] 1-10
                                              10  _CFPreferencesGetAppIntegerValueWithContainer + 40 (CoreFoundation + 349794) [0x7fff20493662] 1-10
                                                10  _CFPreferencesCopyAppValueWithContainerAndConfiguration + 101 (CoreFoundation + 188798) [0x7fff2046c17e] 1-10
                                                  10  -[_CFXPreferences copyAppValueForKey:identifier:container:configurationURL:] + 137 (CoreFoundation + 188996) [0x7fff2046c244] 1-10
                                                    10  -[_CFXPreferences withSearchListForIdentifier:container:cloudConfigurationURL:perform:] + 372 (CoreFoundation + 1600667) [0x7fff205c4c9b] 1-10
                                                      10  __108-[_CFXPreferences(SearchListAdditions) withSearchListForIdentifier:container:cloudConfigurationURL:perform:]_block_invoke + 279 (CoreFoundation + 190322) [0x7fff2046c772] 1-10
                                                        10  -[_CFXPreferences withSearchLists:] + 60 (CoreFoundation + 1600268) [0x7fff205c4b0c] 1-10
                                                          10  __108-[_CFXPreferences(SearchListAdditions) withSearchListForIdentifier:container:cloudConfigurationURL:perform:]_block_invoke.166 + 535 (CoreFoundation + 1601332) [0x7fff205c4f34] 1-10
                                                            10  -[CFPrefsSearchListSource addSourceForIdentifier:user:byHost:container:] + 105 (CoreFoundation + 214433) [0x7fff204725a1] 1-10
                                                              10  -[_CFXPreferences withSourceForIdentifier:user:byHost:container:cloud:perform:] + 532 (CoreFoundation + 1463770) [0x7fff205a35da] 1-10
                                                                10  _CFPrefsCreateCacheKeyForQuintuplet + 68 (CoreFoundation + 217184) [0x7fff20473060] 1-10
                                                                  10  _CFPrefsGetCacheStringForBundleID + 71 (CoreFoundation + 185041) [0x7fff2046b2d1] 1-10
                                                                    10  CFBundleGetMainBundle + 205 (CoreFoundation + 47123) [0x7fff20449813] 1-10
                                                                      10  dlopen + 28 (libdyld.dylib + 22638) [0x7fff203d286e] 1-10
                                                                        10  dlopen_internal(char const*, int, void*) + 123 (libdyld.dylib + 94090) [0x7fff203e3f8a] 1-10
                                                                          10  dyld3::dlopen_internal(char const*, int, void*) + 38 (libdyld.dylib + 22917) [0x7fff203d2985] 1-10
                                                                            10  _pthread_mutex_firstfit_lock_slow + 211 (libsystem_pthread.dylib + 8636) [0x7fff203c31bc] 1-10
                                                                              10  <patched truncated backtrace> 1-10
                                                                                10  __psynch_mutexwait + 10 (libsystem_kernel.dylib + 12486) [0x7fff203950c6] 1-10
                                                                                 *10  psynch_mtxcontinue + 0 (pthread + 9387) [0xffffff80034a34ab] (blocked by turnstile waiting for akd [26407] thread 0xe96f5) 1-10

akd [26407] thread 0xe96f5 blocked by akd [26407] thread 0xe9706:
Thread 0xe96f5    10 samples (1-10)    priority 47 (base 47)
  <thread QoS user initiated (requested default, promote user interactive), process sudden termination dirty, process unclamped, process received importance donation from appMac [27215], IO tier 0>
  10  start + 1 (libdyld.dylib + 87585) [0x7fff203e2621] 1-10
    10  ??? (akd + 153432) [0x10b5ef758] 1-10
      10  -[NSRunLoop(NSRunLoop) run] + 76 (Foundation + 975748) [0x7fff212d9384] 1-10
        10  -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 212 (Foundation + 393121) [0x7fff2124afa1] 1-10
          10  CFRunLoopRunSpecific + 563 (CoreFoundation + 521934) [0x7fff204bd6ce] 1-10
            10  __CFRunLoopRun + 2731 (CoreFoundation + 526418) [0x7fff204be852] 1-10
              10  __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9 (CoreFoundation + 780656) [0x7fff204fc970] 1-10
                10  _dispatch_main_queue_callback_4CF + 613 (libdispatch.dylib + 64063) [0x7fff20228a3f] 1-10
                  10  _dispatch_mach_invoke + 498 (libdispatch.dylib + 112056) [0x7fff202345b8] 1-10
                    10  _dispatch_lane_serial_drain + 263 (libdispatch.dylib + 38055) [0x7fff202224a7] 1-10
                      10  _dispatch_mach_msg_invoke + 441 (libdispatch.dylib + 109127) [0x7fff20233a47] 1-10
                        10  _dispatch_client_callout4 + 9 (libdispatch.dylib + 14439) [0x7fff2021c867] 1-10
                          10  _xpc_connection_mach_event + 935 (libxpc.dylib + 47772) [0x7fff20109a9c] 1-10
                            10  _xpc_connection_call_event_handler + 44 (libxpc.dylib + 52252) [0x7fff2010ac1c] 1-10
                              10  __XPC_CONNECTION_EVENT_HANDLER_CALLOUT__ + 7 (libxpc.dylib + 64773) [0x7fff2010dd05] 1-10
                                10  ??? (akd + 428525) [0x10b6329ed] 1-10
                                  10  ??? (akd + 429806) [0x10b632eee] 1-10
                                    10  ??? (akd + 429262) [0x10b632cce] 1-10
                                      10  -[AKDevice isProtectedWithPasscode] + 50 (AuthKit + 248892) [0x7fff2a368c3c] 1-10
                                        10  SystemAdministrationLibraryCore + 117 (AuthKit + 241499) [0x7fff2a366f5b] 1-10
                                          10  _sl_dlopen_audited + 45 (SoftLinking + 3877) [0x7fff2a5cff25] 1-10
                                            10  dlopen_internal(char const*, int, void*) + 123 (libdyld.dylib + 94090) [0x7fff203e3f8a] 1-10
                                              10  dyld3::dlopen_internal(char const*, int, void*) + 204 (libdyld.dylib + 23083) [0x7fff203d2a2b] 1-10
                                                10  dyld3::AllImages::dlopen(Diagnostics&, char const*, bool, bool, bool, bool, bool, void const*, bool) + 947 (libdyld.dylib + 121577) [0x7fff203eaae9] 1-10
                                                  10  dyld3::AllImages::loadImage(Diagnostics&, char const*, unsigned int, dyld3::closure::DlopenClosure const*, bool, bool, bool, bool, void const*) + 900 (libdyld.dylib + 122934) [0x7fff203eb036] 1-10
                                                    10  dyld3::closure::Image::forEachImageToInitBefore(void (unsigned int, bool&) block_pointer) const + 86 (libdyld.dylib + 19752) [0x7fff203d1d28] 1-10
                                                      10  invocation function for block in dyld3::AllImages::runInitialzersBottomUp(dyld3::closure::Image const*) + 263 (libdyld.dylib + 20059) [0x7fff203d1e5b] 1-10
                                                        10  load_images + 1559 (libobjc.A.dylib + 23017) [0x7fff202639e9] 1-10
                                                          10  +[NSApplication load] + 71 (AppKit + 11719) [0x7fff22ca1dc7] 1-10
                                                            10  +[NSBundle mainBundle] + 71 (Foundation + 13670) [0x7fff211ee566] 1-10
                                                              10  CFBundleGetMainBundle + 26 (CoreFoundation + 46944) [0x7fff20449760] 1-10
                                                                10  _pthread_mutex_firstfit_lock_slow + 211 (libsystem_pthread.dylib + 8636) [0x7fff203c31bc] 1-10
                                                                  10  <patched truncated backtrace> 1-10
                                                                    10  __psynch_mutexwait + 10 (libsystem_kernel.dylib + 12486) [0x7fff203950c6] 1-10
                                                                     *10  psynch_mtxcontinue + 0 (pthread + 9387) [0xffffff80034a34ab] (blocked by turnstile waiting for akd [26407] thread 0xe9706) 1-10
```

akd looks kind of deadlocked on itself from two threads calling out to CFBundleGetMainBundle()

And appMac is blocked by akd when it's doing some authorization stuff which I'm honestly not sure we need.

It makes sense restarting the Mac solves it because akd is a system running process (/System/Library/PrivateFrameworks/AuthKit.framework/Versions/A/Support/akd)

I'd also file a feedback report to Apple if you haven't already.

**Comment by aleksiy805** [comment]
@zorgiepoo thank you for your answer! I've just described the problem on the Apple feedback service and attached this report file. I really think it's a problem in Big Sur, not in my app and of course not in your framework. But on the Internet, I did not find anything even closely similar in description to my problem. Will see, maybe Apple give me feedback for this.

**Comment by aleksiy805** [comment]
Macs with such a problem also have problem with Safari opening. It also becomes "Application Not Responding". I couldn't get crash report because app didn't crash and simply "Application Not Responding". Console.app Log Reports also doesn't show any useful info.

**Comment by zorgiepoo** [comment]
This is an OS bug which I hear might be fixed in the just released macOS 11.5 update and I'm guessing it hasn't been seen in Catalina or earlier. So please see if this gets hit on an updated macOS if you can and if it does, re-open this issue.

**Comment by aleksiy805** [comment]
> This is an OS bug which I hear might be fixed in the just released macOS 11.5 update and I'm guessing it hasn't been seen in Catalina or earlier. So please see if this gets hit on an updated macOS if you can and if it does, re-open this issue.

Ok! Thank you
